pkg/dynamicinstrumentation/diconfig/location_expression.go

```diff
@@ -0,0 +1,322 @@
+					)
+				} else if elementParam.Kind == uint(reflect.Struct) {
+					// Structs don't provide context on location, or have values themselves
+					// but we know that if there's a struct, the next element will have to have
+					// the offset applied
+					if len(pathElements) > pathElementIndex+1 {
+						// Apply the appropriate offset for the next element (the struct field)
+						structField := getParamFromTriePaths(pathElements[pathElementIndex+1])
+						targetExpressions = append(targetExpressions,
+							ditypes.ApplyOffsetLocationExpression(uint(structField.FieldOffset)),
+						)
+					}
+					continue
+				} else if elementParam.Kind == uint(reflect.String) {
+					if len(instrumentationTarget.ParameterPieces) != 2 {
+						continue
+					}
+					stringCharArray := instrumentationTarget.ParameterPieces[0]
+					stringLength := instrumentationTarget.ParameterPieces[1]
+					if stringCharArray == nil || stringLength == nil {
+						continue
+					}
+
+					if stringLength.Location != nil {
+						stringLength.LocationExpressions = append(stringLength.LocationExpressions,
+							ditypes.DirectReadLocationExpression(stringLength),
+							ditypes.PopLocationExpression(1, 2),
+						)
+					} else {
+						stringLength.LocationExpressions = append(targetExpressions,
+							ditypes.ApplyOffsetLocationExpression(uint(stringLength.FieldOffset)),
+							ditypes.DereferenceToOutputLocationExpression(2),
+						)
+					}
+
+					if stringCharArray.Location != nil && stringLength.Location != nil {
+						// Fields of the string are directly assigned
+						targetExpressions = append(targetExpressions,
+							// Read string dynamically:
+							ditypes.DirectReadLocationExpression(stringCharArray),
+							ditypes.DirectReadLocationExpression(stringLength),
+							ditypes.DereferenceDynamicToOutputLocationExpression(uint(limitsInfo.InstrumentationOptions.StringMaxSize)),
+						)
+					} else {
+						// Expect address of the string struct itself on the location expression stack
+						targetExpressions = append(targetExpressions,
+							ditypes.ReadStringToOutput(uint16(limitsInfo.InstrumentationOptions.StringMaxSize)),
+						)
+					}
+					continue
+					/* end parsing strings */
+				} else if elementParam.Kind == uint(reflect.Slice) {
+					if len(elementParam.ParameterPieces) != 3 {
+						continue
+					}
+					sliceIdentifier := randomLabel()
+					slicePointer := elementParam.ParameterPieces[0]
+					sliceLength := elementParam.ParameterPieces[1]
+					if sliceLength.Location != nil {
+						sliceLength.LocationExpressions = append(sliceLength.LocationExpressions,
+							ditypes.DirectReadLocationExpression(sliceLength),
+							ditypes.PopLocationExpression(1, 2),
+						)
+					} else {
+						sliceLength.LocationExpressions = append(targetExpressions,
+							ditypes.ApplyOffsetLocationExpression(uint(sliceLength.FieldOffset)),
+							ditypes.DereferenceToOutputLocationExpression(2),
+						)
+					}
```

Comment:
Is it correct to use targetExpression in the else branch?

If yes, please add a comment explaining why.

Reply:
Wow great catch, thank you!